flow/beta/cli/config/config.go

```diff
@@ -83,33 +160,177 @@ func (k EmulatorServiceKey) MarshalJSON() ([]byte, error) {
-	ChainID flow.ChainID `json:"chain"`
-	Keys    []AccountKey `json:"keys"`
+func (c *NetworkCollection) UnmarshalJSON(b []byte) error {
+	raw := make(map[string]json.RawMessage)
+	c.Networks = make([]Network, 0)
+
+	err := json.Unmarshal(b, &raw)
+	if err != nil {
+		return err
+	}
+
+	for name, v := range raw {
+		network := new(Network)
+		network.Name = name
+
+		err := json.Unmarshal(v, &network)
+		if err != nil {
+			return err
+		}
+
+		c.Networks = append(c.Networks, *network)
+	}
+
+	return nil
 }
 
-type AccountKey struct {
-	Type     KeyType
-	Index    int
-	SigAlgo  crypto.SignatureAlgorithm
-	HashAlgo crypto.HashAlgorithm
-	Context  map[string]string
+//UnmarshalJSON contracts so we can convert it to our structure
+func (c *ContractCollection) UnmarshalJSON(b []byte) error {
+	raw := make(map[string]json.RawMessage)
+	sourceNetwork := make(map[string]string)
+	c.Contracts = make([]Contract, 0)
+
+	err := json.Unmarshal(b, &raw)
+	if err != nil {
+		return err
+	}
+
+	for name, value := range raw {
+		err := json.Unmarshal(value, &sourceNetwork)
+		// advanced schema
+		if err == nil && len(sourceNetwork) > 0 {
+			for network, source := range sourceNetwork {
+				contract := new(Contract)
+				contract.Name = name
+				contract.Network = network
+				contract.Source = source
+				c.Contracts = append(c.Contracts, *contract)
+			}
+		} else { // basic schema
+			contract := new(Contract)
+			contract.Name = name
+			json.Unmarshal(value, &contract.Source)
+			c.Contracts = append(c.Contracts, *contract)
+		}
+	}
+
+	return nil
 }
 
-type accountKeyJSON struct {
-	Type     KeyType           `json:"type"`
-	Index    int               `json:"index"`
-	SigAlgo  string            `json:"signatureAlgorithm"`
-	HashAlgo string            `json:"hashAlgorithm"`
-	Context  map[string]string `json:"context"`
+// UnmarshalJSON account collection to get the key name for the account as well
+func (c *AccountCollection) UnmarshalJSON(b []byte) error {
+	c.Accounts = make(map[string]Account)
+	raw := make(map[string]json.RawMessage)
+
+	err := json.Unmarshal(b, &raw)
+	if err != nil {
+		return err
+	}
+
+	for name, value := range raw {
+		account := new(Account)
+		err := json.Unmarshal(value, &account)
+		if err != nil {
+			return err
+		}
+
+		account.Name = name
+
+		c.Accounts[name] = *account
+	}
+
+	return nil
 }
 
+// UnmarshalJSON Account decodes json config for account
+// and has two options for keys - string and key object
+func (a *Account) UnmarshalJSON(b []byte) error {
+	raw := make(map[string]json.RawMessage)
+
+	err := json.Unmarshal(b, &raw)
+	if err != nil {
+		return err
+	}
+
+	json.Unmarshal(raw["chain"], &a.ChainID)
+
+	var address string
+	json.Unmarshal(raw["address"], &address)
+
+	// TODO: address validation format
+	if address == "service" {
+		if a.ChainID == "" {
+			a.ChainID = flow.Emulator //TODO: find better way for defaults in general
+		}
+
+		a.Address = flow.ServiceAddress(a.ChainID)
```

Comment:
I think it would make sense to add this special logic to the "transform" step in order simplify this `UnmarshalJSON` function.